library/core/src/alloc/global.rs

```diff
@@ -20,24 +20,57 @@ use crate::ptr;
 ///
 /// # Example
 ///
-/// ```no_run
-/// use std::alloc::{GlobalAlloc, Layout, alloc};
+/// ```
+/// use std::alloc::{GlobalAlloc, Layout};
+/// use std::cell::UnsafeCell;
 /// use std::ptr::null_mut;
+/// use std::sync::atomic::{AtomicUsize, Ordering::{Acquire, SeqCst}};
 ///
-/// struct MyAllocator;
-///
-/// unsafe impl GlobalAlloc for MyAllocator {
-///     unsafe fn alloc(&self, _layout: Layout) -> *mut u8 { null_mut() }
-///     unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {}
+/// const ARENA: usize = 128 * 1024;
+/// #[repr(C, align(131072))] // 131072 == ARENA.
+/// struct SimpleAllocator {
+///     arena: UnsafeCell<[u8; ARENA]>,
+///     remaining: AtomicUsize, // we allocate from the top, counting down
 /// }
 ///
 /// #[global_allocator]
-/// static A: MyAllocator = MyAllocator;
+/// static ALLOCATOR: SimpleAllocator = SimpleAllocator {
+///     arena: UnsafeCell::new([0x55; ARENA]),
+///     remaining: AtomicUsize::new(ARENA),
+/// };
 ///
-/// fn main() {
-///     unsafe {
-///         assert!(alloc(Layout::new::<u32>()).is_null())
+/// unsafe impl Sync for SimpleAllocator { }
+///
+/// unsafe impl GlobalAlloc for SimpleAllocator {
+///     unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+///         let size = layout.size();
+///         let align = layout.align();
+///
+///         // Both align==0 and align not power of 2 will mangle remaining (or panic), causing UB.
+///         // This is allowed by the `GlobalAllocator` and `Layout` contracts.
```

Comment:
Actually the `Layout` contract prohibits creating a layout with an invalid alignment: https://doc.rust-lang.org/nightly/std/alloc/struct.Layout.html#method.from_size_align

Reply:
Yes.  That is what I was trying to say.  I have reworded this since evidently it wasn't clear.  (In my updated branch which I will push shortly.)